File: apps/app/lib/notion-oauth.ts
import { randomBytes } from "crypto";

export const NOTION_AUTHORIZE = "https://api.notion.com/v1/oauth/authorize";
export const NOTION_TOKEN = "https://api.notion.com/v1/oauth/token";
export const NOTION_SEARCH = "https://api.notion.com/v1/search";
export const NOTION_VERSION = "2022-06-28";

export function notionClientId(): string | undefined {
  return process.env.NOTION_CLIENT_ID?.trim() || undefined;
}

export function notionClientSecret(): string | undefined {
  return process.env.NOTION_CLIENT_SECRET?.trim() || undefined;
}

export function notionRedirectUri(origin?: string): string {
  const configured = process.env.NOTION_REDIRECT_URI?.trim();
  if (configured) return configured;
  if (origin) return `${origin.replace(/\/$/, "")}/api/auth/notion/callback`;
  return "http://localhost:3001/api/auth/notion/callback";
}

export function generateState(): string {
  return randomBytes(24).toString("hex");
}

export function buildAuthorizeUrl(opts: {
  clientId: string;
  redirectUri: string;
  state: string;
}): string {
  const url = new URL(NOTION_AUTHORIZE);
  url.searchParams.set("client_id", opts.clientId);
  url.searchParams.set("response_type", "code");
  url.searchParams.set("owner", "user");
  url.searchParams.set("redirect_uri", opts.redirectUri);
  url.searchParams.set("state", opts.state);
  return url.toString();
}

export type NotionTokenResponse = {
  access_token: string;
  token_type?: string;
  bot_id?: string;
  workspace_id?: string;
  workspace_name?: string;
  workspace_icon?: string | null;
  duplicated_template_id?: string | null;
};

export async function exchangeCode(opts: {
  code: string;
  clientId: string;
  clientSecret: string;
  redirectUri: string;
}): Promise<NotionTokenResponse> {
  const credentials = Buffer.from(
    `${opts.clientId}:${opts.clientSecret}`,
  ).toString("base64");
  const res = await fetch(NOTION_TOKEN, {
    method: "POST",
    headers: {
      Authorization: `Basic ${credentials}`,
      "Content-Type": "application/json",
      Accept: "application/json",
    },
    body: JSON.stringify({
      grant_type: "authorization_code",
      code: opts.code,
      redirect_uri: opts.redirectUri,
    }),
  });
  const text = await res.text();
  if (!res.ok) {
    throw new Error(`token exchange ${res.status}: ${text.slice(0, 500)}`);
  }
  return JSON.parse(text) as NotionTokenResponse;
}

export type NotionPageSummary = {
  id: string;
  title: string;
  url?: string;
};

function pageTitle(page: Record<string, unknown>): string {
  const props = page.properties;
  if (props && typeof props === "object") {
    for (const value of Object.values(props as Record<string, unknown>)) {
      if (!value || typeof value !== "object") continue;
      const prop = value as Record<string, unknown>;
      if (prop.type === "title" && Array.isArray(prop.title)) {
        const parts = (prop.title as Array<{ plain_text?: string }>)
          .map((t) => t.plain_text || "")
          .join("");
        if (parts) return parts;
      }
    }
  }
  return "Untitled";
}

/** Pages the OAuth bot can see (shared during Notion consent). */
export async function searchPages(
  accessToken: string,
  pageSize = 50,
): Promise<NotionPageSummary[]> {
  const res = await fetch(NOTION_SEARCH, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${accessToken}`,
      "Notion-Version": NOTION_VERSION,
      "Content-Type": "application/json",
      Accept: "application/json",
    },
    body: JSON.stringify({
      filter: { property: "object", value: "page" },
      page_size: pageSize,
    }),
  });
  const text = await res.text();
  if (!res.ok) {
    throw new Error(`notion search ${res.status}: ${text.slice(0, 400)}`);
  }
  const payload = JSON.parse(text) as {
    results?: Array<Record<string, unknown>>;
  };
  const out: NotionPageSummary[] = [];
  for (const page of payload.results || []) {
    if (page.object !== "page" || typeof page.id !== "string") continue;
    out.push({
      id: page.id,
      title: pageTitle(page),
      url: typeof page.url === "string" ? page.url : undefined,
    });
  }
  return out;
}

File: backend/notion_oauth.py
"""Notion helpers for OAuth-connected workspaces (search pages for root picker)."""
from __future__ import annotations

import httpx

NOTION_VERSION = "2022-06-28"
NOTION_SEARCH = "https://api.notion.com/v1/search"


class NotionOAuthError(Exception):
    def __init__(self, message: str, *, status: int | None = None, body: str = ""):
        super().__init__(message)
        self.status = status
        self.body = body


def _headers(access_token: str) -> dict[str, str]:
    return {
        "Authorization": f"Bearer {access_token}",
        "Notion-Version": NOTION_VERSION,
        "Content-Type": "application/json",
    }


def _page_title(page: dict) -> str:
    props = page.get("properties") or {}
    if isinstance(props, dict):
        for value in props.values():
            if not isinstance(value, dict):
                continue
            if value.get("type") == "title" and isinstance(value.get("title"), list):
                parts = "".join(
                    (t.get("plain_text") or "") for t in value["title"] if isinstance(t, dict)
                )
                if parts:
                    return parts
    return "Untitled"


def search_pages(access_token: str, *, page_size: int = 50) -> list[dict]:
    with httpx.Client(timeout=30.0) as client:
        res = client.post(
            NOTION_SEARCH,
            headers=_headers(access_token),
            json={
                "filter": {"property": "object", "value": "page"},
                "page_size": page_size,
            },
        )
    if res.status_code >= 400:
        raise NotionOAuthError(
            "Failed to search Notion pages",
            status=res.status_code,
            body=res.text[:500],
        )
    payload = res.json()
    out: list[dict] = []
    for page in payload.get("results") or []:
        if not isinstance(page, dict) or page.get("object") != "page":
            continue
        pid = page.get("id")
        if not pid:
            continue
        out.append(
            {
                "id": pid,
                "title": _page_title(page),
                "url": page.get("url"),
            }
        )
    return out
